### PR TITLE
Wire up frontend

### PR DESCRIPTION
**Pull request by yimothysu**
This PR wires up the frontend to the working backend for an end-to-end OpenDevin.

[Screencast from 03-24-2024 08:39:30 PM.webm](https://github.com/OpenDevin/OpenDevin/assets/46636043/8f9ee289-6299-4b4c-8aba-8c82bb0363e3)

### CODE REVIEW

File: frontend/src/components/Terminal.tsx [hunk 1]
```diff
@@ -22,9 +22,15 @@ class JsonWebsocketAddon {
...
-      if (message.action === "terminal") {
-        terminal.write(message.data);
+      const { action, args } = JSON.parse(event.data);
+      if (action === "run") {
+        terminal.writeln(args.command);
+      }
+      if (action === "output") {
+        args.output.split("\n").forEach((line: string) => {
+          terminal.writeln(line);
+        });
```
Comment (rbren) [inline frontend/src/components/Terminal.tsx:32]: in the future, it might be nice to navigate to this file in the code editor pane

**Review by rbren — APPROVED** [review]
This is awesome!

**Comment by xcodebuild** [comment]
Thanks for your awesome work. It seems we need to start the backend to develop the frontend project now, so wouldn't it be best to add some tips on how to start the relevant services?

It would be significantly beneficial if we could simulate the backend service, making front-end development engagement more straightforward.

**Comment by huybery** [comment]
LGTM

**Comment by rbren** [comment]
@yimothysu I'm actually having trouble getting the FE to talk to the BE--the `/ws` request just hangs as pending. I've tried to run backend listening on :3000 and :8000. Any ideas?

**Comment by rbren** [comment]
Got it working--I had to change the port to `:8000` in .env. Not sure why `:3000` wasn't working (and I'm surprised my machine let me start a websocket server on 3000 and the frontend on 3000 at the same time 🤔)

File: frontend/src/components/Terminal.tsx (hunk 1 above)
Comment (yimothysu) [reply frontend/src/components/Terminal.tsx:32]: good point!

**Comment by yimothysu** [comment]
> Thanks for your awesome work. It seems we need to start the backend to develop the frontend project now, so wouldn't it be best to add some tips on how to start the relevant services?
> 
> It would be significantly beneficial if we could mock the backend service, making front-end development engagement more straightforward.

@xcodebuild Good idea! I think we need to streamline some READMEs to clarify how to run the service end-to-end. I can do this.

A backend mock would definitely be useful. How do you suggest we handle this?

**Pull request merged**